frame/assets/src/functions.rs

```diff
@@ -45,34 +61,41 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	pub(super) fn new_account(
 		who: &T::AccountId,
 		d: &mut AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>,
-	) -> Result<bool, DispatchError> {
+		maybe_deposit: Option<DepositBalanceOf<T, I>>,
+	) -> Result<ExistenceReason<DepositBalanceOf<T, I>>, DispatchError> {
 		let accounts = d.accounts.checked_add(1).ok_or(ArithmeticError::Overflow)?;
-		let is_sufficient = if d.is_sufficient {
+		let reason = if let Some(deposit) = maybe_deposit {
+			ExistenceReason::DepositHeld(deposit)
+		} else if d.is_sufficient {
 			frame_system::Pallet::<T>::inc_sufficients(who);
 			d.sufficients += 1;
-			true
+			ExistenceReason::Sufficient
 		} else {
 			frame_system::Pallet::<T>::inc_consumers(who).map_err(|_| Error::<T, I>::NoProvider)?;
-			false
+			ExistenceReason::Consumer
 		};
 		d.accounts = accounts;
-		Ok(is_sufficient)
+		Ok(reason)
 	}
 
 	pub(super) fn dead_account(
 		what: T::AssetId,
 		who: &T::AccountId,
 		d: &mut AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>,
-		sufficient: bool,
-	) {
-		if sufficient {
-			d.sufficients = d.sufficients.saturating_sub(1);
-			frame_system::Pallet::<T>::dec_sufficients(who);
-		} else {
-			frame_system::Pallet::<T>::dec_consumers(who);
+		reason: &ExistenceReason<DepositBalanceOf<T, I>>,
+	) -> DeadConsequence {
```

Comment:
in `do_destroy` it seems the result of `dead_account` is not handled correctly, if some account are still existing because of deposit held, then they are not removed, but the asset is removed.

Reply:
There is a comment by its usage:

```
// We have to force this as it's destroying the entire asset class.
// This could mean that some accounts now have irreversibly reserved
// funds.
let _ = Self::dead_account(id, &who, &mut details, &v.reason, true);
```